internal/ethapi/transaction_args.go

```diff
@@ -141,7 +141,7 @@ func (args *TransactionArgs) setFeeDefaults(ctx context.Context, b Backend) erro
 	// who are not yet synced past London to get defaults for other tx values. See
 	// https://github.com/ethereum/go-ethereum/pull/23274 for more information.
 	eip1559ParamsSet := args.MaxFeePerGas != nil && args.MaxPriorityFeePerGas != nil
-	if (args.GasPrice != nil && !eip1559ParamsSet) || (args.GasPrice == nil && eip1559ParamsSet) {
+	if (args.GasPrice != nil && args.GasPrice.ToInt().Sign() > 0 && !eip1559ParamsSet) || (args.GasPrice == nil && eip1559ParamsSet) {
```

Comment:
`args.GasPrice` is very likely `nil` at this point

Reply:
I think maybe this clause can be split up, before more stuff is added. Something like this:
```golang
	if args.GasPrice == nil && eip1559ParamsSet {
		// Sanity check the EIP-1559 fee parameters if present.
		if args.MaxFeePerGas.ToInt().Cmp(args.MaxPriorityFeePerGas.ToInt()) < 0 {
			return fmt.Errorf("maxFeePerGas (%v) < maxPriorityFeePerGas (%v)", args.MaxFeePerGas, args.MaxPriorityFeePerGas)
		}
		return nil // No need to set anything, user already set MaxFeePerGas and MaxPriorityFeePerGas
	}
	if args.GasPrice != nil && !eip1559ParamsSet {
		return nil // No need to set anything, user already set GasPrice
	}
```

Reply:
> But what about 1-priced post-london txs

AFAIK, if the user doesn't give the `gasLimit`, it will be checked in `estimateGas` in https://github.com/ethereum/go-ethereum/blob/afd421766c132fb0440d94f6978418b7a3709944/core/state_transition.go#L310

If the `gasLimit` is given, there seems no checking. if there is any spec of base gasPrice

Reply:
> The setDefaults should fill the blanks

Seems there is too much validation logic in the `setDefaults`, maybe we need to rename this function to `checkAndSetDefaults` or something else.